src/lib/message/L2ToL1Message.ts

```diff
@@ -345,80 +327,88 @@ export class L2ToL1MessageReader extends L2ToL1Message {
 
-    // TODO: use IRollupUser interface instead
-    const rollup = RollupUserFacet__factory.connect(
-      network.ethBridge.rollup,
+    const rollup = RollupUserLogic__factory.connect(
+      l2Network.ethBridge.rollup,
       this.l1Provider
     )
 
-    const proof = await this.tryGetProof(l2Provider)
-    // here we assume the L2 to L1 tx is actually valid, so the user needs to wait the max time.
-    if (proof === null)
-      return BigNumber.from(network.confirmPeriodBlocks)
-        .add(ASSERTION_CREATED_PADDING)
-        .add(ASSERTION_CONFIRMED_PADDING)
-    // we can't check if the L2 to L1 tx isSpent on the outbox, so we instead try executing it
-    if (await this.hasExecuted(proof)) return BigNumber.from(0)
+    const status = await this.status(l2Provider)
+    if (status === L2ToL1MessageStatus.EXECUTED) return null
+    if (status === L2ToL1MessageStatus.CONFIRMED) {
+      const latestConfirmed = await rollup.callStatic.latestConfirmed()
+      const node = await rollup.getNode(latestConfirmed)
+      return node.deadlineBlock
+    }
+    if (status === L2ToL1MessageStatus.NOT_FOUND)
+      throw new ArbTsError('L2ToL1Msg not found')
+
+    // consistency check in case we change the enum in the future
+    if (status !== L2ToL1MessageStatus.UNCONFIRMED)
+      throw new ArbTsError('L2ToL1Msg expected to be unconfirmed')
+
     const latestBlock = await this.l1Provider.getBlockNumber()
 
     const eventFetcher = new EventFetcher(this.l1Provider)
 
-    const events = (
+    const logs = (
       await eventFetcher.getEvents(
-        network.ethBridge.rollup,
-        RollupUserFacet__factory,
+        rollup.address,
+        RollupUserLogic__factory,
         t => t.filters.NodeCreated(),
         {
-          // ~40k blocks with a 15sec blocktime and 8days confirmPeriodBlocks
-          fromBlock:
+          fromBlock: Math.max(
             latestBlock -
-            BigNumber.from(network.confirmPeriodBlocks)
-              .add(ASSERTION_CONFIRMED_PADDING)
-              .toNumber(),
-          toBlock: latestBlock,
+              BigNumber.from(l2Network.confirmPeriodBlocks)
+                .add(ASSERTION_CONFIRMED_PADDING)
+                .toNumber(),
+            0
+          ),
+          toBlock: 'latest',
         }
       )
-    )
-      .map(e => e.event)
-      .filter(e => {
-        const afterSendCount = e.assertionIntFields[1][2]
-        return BigNumber.from(afterSendCount).gte(this.batchNumber)
-      })
-      .sort((a, b) => {
-        return (
-          BigNumber.from(a.assertionIntFields[1][2]).toNumber() -
-          BigNumber.from(b.assertionIntFields[1][2]).toNumber()
-        )
-      })
-
-    // a node that covers this tx still has not been created
-    if (events.length === 0)
-      return BigNumber.from(network.confirmPeriodBlocks)
+    ).sort((a, b) => a.event.nodeNum.toNumber() - b.event.nodeNum.toNumber())
+
+    let found = false
+    let logIndex = 0
+    while (!found) {
```

Comment:
Should use a for loop over `logs` instead

Reply:
made it a while loop to make it into a binary search in the future to optimise this